library/ecp.c

```diff
@@ -2061,6 +2061,57 @@ int mbedtls_ecp_check_pub_priv( const mbedtls_ecp_keypair *pub, const mbedtls_ec
     return( ret );
 }
 
+#if defined(MBEDTLS_ASN1_WRITE_C) && defined(MBEDTLS_OID_C)
+#include "mbedtls/asn1write.h"
+#include "mbedtls/oid.h"
+int mbedtls_ecp_ansi_write_group( const mbedtls_ecp_group *grp,
+                                  unsigned char *p,
+                                  size_t size )
+{
+    const char *oid;
+    unsigned char *q;
+    size_t oid_length;
+    int ret;
+    ret = mbedtls_oid_get_oid_by_ec_grp( grp->id, &oid, &oid_length );
+    if( ret != 0 )
+        return( ret );
+    // Output is a TLV with len(T)=1, len(L)=1, V=OID
+    if( size < 2 + oid_length )
+        return( MBEDTLS_ERR_ECP_BUFFER_TOO_SMALL );
+    q = p + 2 + oid_length;
+    return( mbedtls_asn1_write_oid( &q, p, oid, oid_length ) );
```

Comment:
Suggestion: Assert that `q == p` in the end to become aware of changes in the size of encodings.

Reply:
Shouldn't that be `q >= p`? AFAIR this function doesn't require an exact buffer size.

Reply:
But we do want the content to be written starting from the beginning of the buffer, so I think we should either check that `p == q` or include a `memmove` as in `mbedtls_ecp_ansi_write_point`.

Reply:
What kind of an error would you consider suitable for a case when p != q? MBEDTLS_ERR_ECP_BUFFER_TOO_SMALL, MBEDTLS_ERR_ECP_BAD_INPUT_DATA, or maybe we treat it as a programmers mistake and only assert it?

Reply:
@gilles-peskine-arm Ping. 

I would prefer to have this function work like `ecp_ansi_write_point`, `memmov`'ing the post-fix of the buffer that has been written to the front.

Reply:
@hanno-arm Sorry, I'm not following. This function does write to the beginning of the buffer, why would there by anything to `memmove`?

The function works correctly assuming that `mbedtls_oid_get_oid_by_ec_grp` and `mbedtls_asn1_write_oid` are consistent. If they're not, then I suppose we should return `MBEDTLS_ERR_ECP_BAD_INPUT_DATA` — treat it as a strange OID. (For example it might fail with an OID that's encoded as more than about 126 bytes, but there's no such OID now or in any likely future.)

Reply:
Added a check for p!=q, returning MBEDTLS_ERR_ECP_BAD_INPUT_DATA in such case.